### ISSUE TITLE
"address already in use", how to change settings on Ubuntu host?

### ISSUE THREAD

**Issue by rucksman** [MatthewVance/unbound-docker#88]
When running the container with the standard command (no changes in config files), I get this error:
`docker: Error response from daemon: driver failed programming external connectivity on endpoint my-unbound (....): Error starting userland proxy: listen tcp4 0.0.0.0:53: bind: address already in use.
`
This is most likely due to systemd-resolved running on the host (Ubuntu 18.04).

What would I have to change on my host so that Unbound is going to work? I probably would have to stop and disable systemd-resolved and also remove the symlink from /etc/resolv.conf to /run/systemd/resolve/stub-resolv.conf. But what next? Create a new resolv.conf file? How should this file look like?

Thanks for helping!

**Comment by MatthewVance** [comment]
Hi @rucksman,

I'll start by saying I'm not a fan of how Ubuntu has implemented networking features. At least for me, netplan and systemd-resolved seem to unnecessarily complicate things. I'm sure they're great for users who like to stick with the defaults and admins who've taken the time to learn the new tools.

You are correct that you will have to stop the service on your host that is using that port so that the Docker container can bind to port 53. Once port 53 is available for use on your host, the container will run with the standard command. 

The other option is to have the container serve DNS on a different port to avoid the conflict and tell your clients to use it: 

```docker run \
--name=my-unbound \
--publish=5353:53/tcp \
--publish=5353:53/udp \
--restart=unless-stopped -\
-detach=true \
mvance/unbound:1.14.0
```

You can check that the Unbound container using a different port works via a dig command: `dig @x.x.x.x -p 5353 google.com` (replacing x.x.x.x with the IP of your host).

On my host running the Docker Unbound container, I let it use a different DNS server rather than try to get it to use the Unbound container running on the same host. Otherwise, a scenario can come up where the host is up and the container is down. This leaves the host with DNS service, unless you have a fallback. 

To be transparent, I've ran into various weird behavior with trying to get Ubuntu DNS to do what I want while also running DNS in a container on port 53 so I've resigned myself to live with some workarounds. My setup is fairly weird so you may have better luck. 

The DNS settings on the KVM host and the VMs are all different and have changed over the years as I've experimented. What I can tell you about my current set up is: 

- My main host runs a few a couple of different VMs via KVM that have Docker installed. This KVM host currently uses 127.0.0.53#53 for DNS; however, this main host never runs Unbound as a container directly (I keep Docker in VMs instead of the main host for a few reasons). 
- One of my VMs that sometimes runs an Unbound container is configured to use DNS running on another host on my LAN that runs an Unbound instance (the other host is a Raspbery Pi). 
- On another VM that sometimes runs an Unbound container, I use the alternate port method mentioned above because this VM is still set to use 127.0.0.53#53 (i.e., systemd-resolved). 
- On my router/gateway, I don't currently have it setup to use any of my Unbound instances or to tell clients relying on DHCP to use them either (instead, I manually configure the DNS settings on the clients/endpoints on my LAN to use my various DNS containers).

I realize all of the above is weird and less than ideal, but it mostly does what I need it to do. My main concern is ensuring certain LAN clients use my Unbound containers. I let the others get the DNS settings I've setup for them to receive via DHCP unless I have a reason to configure them manually to use something else that's not one of my Unbound containers.

All that is to say, I don't have a great baseline setting to lookup and give you. 

Like I started by saying, networking/DNS on Ubuntu is strange. The following links may help:

- https://ubuntu.com/server/docs/network-configuration
- https://help.ubuntu.com/community/Network
- http://manpages.ubuntu.com/manpages/bionic/man8/systemd-resolved.service.8.html
- https://askubuntu.com/questions/907246/how-to-disable-systemd-resolved-in-ubuntu

I'd be curious to know what you settle on for your particular setup.

**Comment by bonelifer** [comment]
Not sure if this will help but from pihole.net:
https://github.com/pi-hole/docker-pi-hole#installing-on-ubuntu
tl;dr
```
sudo sed -r -i.orig 's/#?DNSStubListener=yes/DNSStubListener=no/g' /etc/systemd/resolved.conf
sudo sh -c 'rm /etc/resolv.conf && ln -s /run/systemd/resolve/resolv.conf /etc/resolv.conf'
sudo systemctl restart systemd-resolved
```

**Comment by rucksman** [comment]
@MatthewVance 
Sorry, forgot to answer your very detailed reply (thanks for that!). For different reasons I ended up with creating my own Dockerfile/Docker image (not as advanced as yours, but works for me).
